src/java.base/share/classes/jdk/internal/lang/StableValue.java

```diff
@@ -0,0 +1,325 @@
+     * optimizations by the JVM and is equivalent to:
+     * {@snippet lang = java:
+     * Map<K, StableValue<V>> map = Map.copyOf(keys.stream()
+     *         .distinct()
+     *         .map(Objects::requireNonNull)
+     *         .collect(Collectors.toMap(Function.identity(), _ -> StableValue.of())));
+     * }
+     * Except it requires less storage, does not return stable value instances with the
+     * same identity, and is likely to exhibit better performance.
+     * <p>
+     * This static factory methods return map instances (and with all their values)
+     * that are <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>,
+     * immutable and thread-safe. Programmers should not use map and value instances for
+     * synchronization, or unpredictable behavior may occur. For example, in a
+     * future release, synchronization may fail. Consequently,
+     * the query methods {@linkplain Map#containsValue(Object)} and methods that relies
+     * on this and similar method will always indicate no match.
+     * <p>
+     * As an emerging property, providing an {@linkplain EnumSet} as {@code keys} will
+     * make the returned Map eligible for certain additional optimizations.
+     *
+     * @param keys the keys in the map
+     * @param <K>  the type of keys maintained by the returned map
+     * @param <V>  the type of mapped StableValue values
+     */
+    static <K, V> Map<K, StableValue<V>> ofMap(Set<? extends K> keys) {
+        Objects.requireNonNull(keys);
+        return StableValueImpl.ofMap(keys);
+    }
+
+    /**
+     * If no value {@linkplain #isSet() is set} for the StableValue at the provided
+     * {@code index}, attempts to compute and set it as per
+     * {@linkplain StableValue#computeIfUnset(Supplier)} by applying the provided
+     * {@code mapper}.
+     * <p>
+     * This is equivalent to:
+     * {@snippet lang = java:
+     * StableValue<V> stable = list.get(index);
+     * if (stable.isSet()) {
+     *     return stable.orThrow();
+     * }
+     * Supplier<V> supplier = () -> mapper.apply(index);
+     * return stable.computeIfUnset(supplier);
+     *}
+     * Except it might be more resource efficient and performant.
+     *
+     * @param list   from which to get a StableValue
+     * @param index  for the StableValue
+     * @param mapper to apply if the StableValue at the provided {@code index} is
+     *               {@linkplain StableValue#isSet() not set}
+     * @return the current (pre-existing or computed) value at the provided {@code index}
+     * @param <V> the StableValue type to set
+     * @throws IndexOutOfBoundsException if the provided {@code index} is less than
+     *         zero or {@code index >= list.size()}
+     */
+    static <V> V computeIfUnset(List<StableValue<V>> list,
+                                int index,
+                                IntFunction<? extends V> mapper) {
```

Comment:
Hmm, these APIs seem unintuitive and error-prone to users. Have you studied the use case where for one stable list/map, there are distinct initialization logics for different indices/keys so that they support different mappers for the same list/map? I cannot recall on top of my head. 

If we drop said ability and restrict mappers to the list/map creation, the whole thing will be much cleaner, and it's a better way to avoid capturing lambdas as well. Users can still go to individual stable values and use functional creation if they really, really want that functionality.

Reply:
I agree that these method appear to be confusing. We have:

```
StableValue::of()
StableValue::ofList(int)
StableValue::ofMap(Set)
```

These methods are clearly primitives.

Everything else falls in the "helper" bucket. That is, we could have:

```
StableValue::ofList(IntFunction) -> List<V> // similar to StableValue::ofList(int)
StableValue::ofMap(Function) -> Map<K, V> // similar to StableValue::ofMap(Set)
```

Or, we could have:

```java
StableValue::ofSupplier(Supplier) -> Supplier<V> // similar to StableValue::of()
StableValue::ofIntFunction(IntFunction) -> IntFunction<V> // similar to StableValue::ofList(int)
StableValue::ofFunction(Function) -> Function<K, V> // similar to StableValue::ofMap(Set)
```

IMHO, having both sets feel slightly redundant. That is, if you have a Map<K, V>, you also have a function from K, V - namely, map::get. And, conversely, if a client wants a List<V> of fixed size, which is populated lazily, using a memoized IntFunction is, effectively, the same thing.

Reply:
I prefer these:
```
StableValue::ofSupplier(Supplier<V>) -> StableValue<V>
StableValue::ofIntFunction(keys, IntFunction<V>) -> IntFunction<StableValue<V>>
StableValue::ofFunction(keys, Function<K, V>) -> Function<K, StableValue<V>>
```
These still expose StableValue so users can set the values if they need. In addition, the List/Map functionalites are mostly useless so a getter/function suffices for the most part.

These APIs are less error-prone to accidental context capture compared to the individual use-site ones, as use-site leaks means each access involves an allocation, but the allocation for construction site is shared.